src/cryptography.ts

```diff
@@ -0,0 +1,176 @@
+import { hmac as nobleHmac } from '@noble/hashes/hmac';
+import { pbkdf2Async as noblePbkdf2 } from '@noble/hashes/pbkdf2';
+import { ripemd160 as nobleRipemd160 } from '@noble/hashes/ripemd160';
+import { sha256 as nobleSha256 } from '@noble/hashes/sha256';
+import { keccak_256 as nobleKeccak256 } from '@noble/hashes/sha3';
+import { sha512 as nobleSha512 } from '@noble/hashes/sha512';
+
+import { isWebCryptoSupported } from './utils';
+
+export type CryptographicFunctions = {
+  /**
+   * Compute the HMAC-SHA-512 of the given data using the given key.
+   *
+   * @param key - The key to use.
+   * @param data - The data to hash.
+   * @returns The HMAC-SHA-512 of the data.
+   */
+  hmacSha512?: (key: Uint8Array, data: Uint8Array) => Promise<Uint8Array>;
+
+  /**
+   * Compute the PBKDF2 of the given password, salt, iterations, and key length.
+   * The hash function used is SHA-512.
+   *
+   * @param password - The password to hash.
+   * @param salt - The salt to use.
+   * @param iterations - The number of iterations.
+   * @param keyLength - The desired key length.
+   * @returns The PBKDF2 of the password.
+   */
+  pbkdf2Sha512?: (
+    password: Uint8Array,
+    salt: Uint8Array,
+    iterations: number,
+    keyLength: number,
+  ) => Promise<Uint8Array>;
+};
+
+/**
+ * Compute the HMAC-SHA-512 of the given data using the given key.
+ *
+ * This function uses the Web Crypto API if available, falling back to a
+ * JavaScript implementation if not.
+ *
+ * @param key - The key to use.
+ * @param data - The data to hash.
+ * @param cryptographicFunctions - The cryptographic functions to use. If
+ * provided, these will be used instead of the built-in implementations.
+ * @returns The HMAC-SHA-512 of the data.
+ */
+export async function hmacSha512(
+  key: Uint8Array,
+  data: Uint8Array,
+  cryptographicFunctions: CryptographicFunctions = {},
+): Promise<Uint8Array> {
+  if (cryptographicFunctions.hmacSha512) {
+    return await cryptographicFunctions.hmacSha512(key, data);
+  }
+
+  if (isWebCryptoSupported()) {
+    /* eslint-disable no-restricted-globals */
+    const subtleKey = await crypto.subtle.importKey(
```

Comment:
If possible we should test the two implementations against each other and compare the result